<h1> 앙상블 - 부스팅(Xgboost)

데이터 : 금융데이터. 텔레마케팅시 고객이 예금에 가입하는 여부

그렇다면 어떤 특징을 가진 고객에게 집중해야할까?

In [3]:
import os
import pandas as pd

In [4]:
os.chdir(r'C:\Users\hjb38\Documents\데이터 분석 과정\data\ml_data')
data = pd.read_csv('bank-additional-full.csv', sep = ";")
# 일반적으로 csv 파일의 구분인자는 ,인데
# 이 파일의 구분인자는 ;라서 이렇게 표시해줘야 함.

In [5]:
data.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

<h2> 1. 원핫인코딩으로 범주형을 숫자형으로 변환

In [6]:
# 먼저 dtypes 를 이용해 변수의 특징을 확인한다.
data.dtypes

# object 가 범주형.

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

In [7]:
# 원핫 인코딩 실시. y는 타겟변수로 쓸거라 제외
data = pd.get_dummies(data, columns = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome'])
data.head()

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0   56       261         1    999         0           1.1          93.994   
1   57       149         1    999         0           1.1          93.994   
2   37       226         1    999         0           1.1          93.994   
3   40       151         1    999         0           1.1          93.994   
4   56       307         1    999         0           1.1          93.994   

   cons.conf.idx  euribor3m  nr.employed  ... month_oct  month_sep  \
0          -36.4      4.857       5191.0  ...         0          0   
1          -36.4      4.857       5191.0  ...         0          0   
2          -36.4      4.857       5191.0  ...         0          0   
3          -36.4      4.857       5191.0  ...         0          0   
4          -36.4      4.857       5191.0  ...         0          0   

   day_of_week_fri  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0                0                1                0                0   
1                0                1                0                0   
2                0                1                0                0   
3                0                1                0                0   
4                0                1                0                0   

   day_of_week_wed  poutcome_failure  poutcome_nonexistent  poutcome_success  
0                0                 0                     1                 0  
1                0                 0                     1                 0  
2                0                 0                     1                 0  
3                0                 0                     1                 0  
4                0                 0                     1                 0  

[5 rows x 64 columns]

<h2> 2. train 과 test 데이터로 나누기

In [8]:
# 데이터의 전체 수 확인
data['id'] = range(len(data))

In [9]:
len(data)

41188

In [10]:
# 41188 개의 데이터 중 30000개를 train 데이터로 지정
train = data.sample(30000, replace = False, random_state = 2020).reset_index().drop(['index'], axis = 1)
# 나머지를 test 데이터로
test = data.loc[ ~data['id'].isin(train['id'])].reset_index().drop(['index'], axis = 1)

In [11]:
# 인풋변수로 쓸 train 컬럼들을 확인. y, id 빼고 전부
train.columns

Index(['age', 'duration', 'campaign', 'pdays', 'previous', 'emp.var.rate',
       'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed', 'y',
       'job_admin.', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid',
       'job_management', 'job_retired', 'job_self-employed', 'job_services',
       'job_student', 'job_technician', 'job_unemployed', 'job_unknown',
       'marital_divorced', 'marital_married', 'marital_single',
       'marital_unknown', 'education_basic.4y', 'education_basic.6y',
       'education_basic.9y', 'education_high.school', 'education_illiterate',
       'education_professional.course', 'education_university.degree',
       'education_unknown', 'default_no', 'default_unknown', 'default_yes',
       'housing_no', 'housing_unknown', 'housing_yes', 'loan_no',
       'loan_unknown', 'loan_yes', 'contact_cellular', 'contact_telephone',
       'month_apr', 'month_aug', 'month_dec', 'month_jul', 'month_jun',
       'month_mar', 'month_may', 'month_nov', 'mon

In [12]:
# 너무 기니까 input_var라는 변수에 담아서 사용

input_var = ['age', 'duration', 'campaign', 'pdays', 'previous', 'emp.var.rate',
       'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed', 
       'job_admin.', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid',
       'job_management', 'job_retired', 'job_self-employed', 'job_services',
       'job_student', 'job_technician', 'job_unemployed', 'job_unknown',
       'marital_divorced', 'marital_married', 'marital_single',
       'marital_unknown', 'education_basic.4y', 'education_basic.6y',
       'education_basic.9y', 'education_high.school', 'education_illiterate',
       'education_professional.course', 'education_university.degree',
       'education_unknown', 'default_no', 'default_unknown', 'default_yes',
       'housing_no', 'housing_unknown', 'housing_yes', 'loan_no',
       'loan_unknown', 'loan_yes', 'contact_cellular', 'contact_telephone',
       'month_apr', 'month_aug', 'month_dec', 'month_jul', 'month_jun',
       'month_mar', 'month_may', 'month_nov', 'month_oct', 'month_sep',
       'day_of_week_fri', 'day_of_week_mon', 'day_of_week_thu',
       'day_of_week_tue', 'day_of_week_wed', 'poutcome_failure',
       'poutcome_nonexistent', 'poutcome_success']

In [13]:
# 인풋변수로 쓸 train 컬럼들을 확인. y, id 빼고 전부
train.columns

# 너무 기니까 input_var라는 변수에 담아서 사용

input_var = ['age', 'duration', 'campaign', 'pdays', 'previous', 'emp.var.rate',
       'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed', 
       'job_admin.', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid',
       'job_management', 'job_retired', 'job_self-employed', 'job_services',
       'job_student', 'job_technician', 'job_unemployed', 'job_unknown',
       'marital_divorced', 'marital_married', 'marital_single',
       'marital_unknown', 'education_basic.4y', 'education_basic.6y',
       'education_basic.9y', 'education_high.school', 'education_illiterate',
       'education_professional.course', 'education_university.degree',
       'education_unknown', 'default_no', 'default_unknown', 'default_yes',
       'housing_no', 'housing_unknown', 'housing_yes', 'loan_no',
       'loan_unknown', 'loan_yes', 'contact_cellular', 'contact_telephone',
       'month_apr', 'month_aug', 'month_dec', 'month_jul', 'month_jun',
       'month_mar', 'month_may', 'month_nov', 'month_oct', 'month_sep',
       'day_of_week_fri', 'day_of_week_mon', 'day_of_week_thu',
       'day_of_week_tue', 'day_of_week_wed', 'poutcome_failure',
       'poutcome_nonexistent', 'poutcome_success']

<h1> Xgboost 실습

<h3> 특징

1. 해석이 어렵다.
2. 랜덤 포레스트에 비해 빠르고 성능이 좋은 경우가 많다.

<h3> 파라미터

1. n_estimators : 몇 개의 의사결정나무를 만들 것인지. 보통 100~500개
2. learning_rate : 얼마나 바르게 학습할 것인지. 보통 0.05 ~ 0.1

<h2> 1) Xgboost 학습시키기

In [14]:
# xgboost는 아나콘다에 내장되어있기 않기 때문에 따로 설치해줘야 한다.
!pip install xgboost

In [15]:
from xgboost import XGBClassifier

In [16]:
xgb = XGBClassifier(n_estimators = 300, learning_rate = 0.1)

In [17]:
# train 학습
xgb.fit(train[input_var], train['y'])

C:\ProgramData\Anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[10:34:14] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.1, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=300, n_jobs=16, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [18]:
# test 예측
predictions = xgb.predict(test[input_var])

In [19]:
test['pred'] = predictions

In [20]:
(test['pred'] == test['y']).mean()

0.9141043975688238

91.4% 정확도. 배깅(램덤 포레스트) 보다 0.3% 높다.

<h2> 2) 최적의 n_estimators 찾기

In [34]:
for n in [100, 200, 300, 400, 500, 600, 700, 800, 900]:
    xgb = XGBClassifier(n_estimators = 400, learning_rate = 0.1)
    xgb.fit(train[input_var], train['y'])
    predictions = xgb.predict(test[input_var])
    test['pred'] = predictions
    (test['pred'] == test['y']).mean()
    
# n 을 한개씩 넣으면 경고문은 뜨더라도 결과는 나오는데
# for 문 쓰니까 결과가 안 나옴ㅠㅠ

[15:25:20] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:25:23] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:25:26] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:25:30] WARNING: C:/Users/Administrator/workspace/xgboost-win64_rel

<h2> 2) Xgboost 변수중요도 뽑아내기

In [21]:
# 각 변수에 대한 변수중요도 확인하기(input_var의 순서와 같음)
feature_imp = xgb.feature_importances_
feature_imp

array([0.00793777, 0.04038796, 0.00718578, 0.023131  , 0.00716699,
       0.03359416, 0.01322974, 0.02985962, 0.01581056, 0.4127995 ,
       0.00624862, 0.01139711, 0.00685038, 0.00577246, 0.00625187,
       0.0059287 , 0.00680625, 0.00731968, 0.00561962, 0.00690037,
       0.00718489, 0.00676124, 0.0067717 , 0.00676973, 0.00742636,
       0.00372667, 0.0102065 , 0.00666046, 0.00845055, 0.00824915,
       0.        , 0.00793599, 0.00956898, 0.0072871 , 0.01564412,
       0.        , 0.        , 0.00528927, 0.00566373, 0.00567542,
       0.00856543, 0.        , 0.00855656, 0.01166364, 0.        ,
       0.01039344, 0.00746287, 0.0045788 , 0.01026478, 0.0064241 ,
       0.00546951, 0.02257893, 0.00786941, 0.02590525, 0.0023061 ,
       0.0077565 , 0.00923624, 0.00745689, 0.00741677, 0.00881127,
       0.0107907 , 0.        , 0.02702291], dtype=float32)

In [22]:
# 보기 쉽게 데이터프레임으로 만들기
imp_df = pd.DataFrame({'var':input_var,
                       'imp':feature_imp})

imp_df

var       imp
0                    age  0.007938
1               duration  0.040388
2               campaign  0.007186
3                  pdays  0.023131
4               previous  0.007167
..                   ...       ...
58       day_of_week_tue  0.007417
59       day_of_week_wed  0.008811
60      poutcome_failure  0.010791
61  poutcome_nonexistent  0.000000
62      poutcome_success  0.027023

[63 rows x 2 columns]

In [23]:
# 큰 값부터 순서대로 정렬
imp_df.sort_values(['imp'], ascending=False)
# nr.employed가 결과에 가장 영향 많이 끼침
# 랜덤포레스트와 변수중요도가 다르다? -> Xgboost의 변수중요도가 더 신뢰도가 높다!

var       imp
9            nr.employed  0.412800
1               duration  0.040388
5           emp.var.rate  0.033594
7          cons.conf.idx  0.029860
62      poutcome_success  0.027023
..                   ...       ...
36           default_yes  0.000000
35       default_unknown  0.000000
30  education_illiterate  0.000000
61  poutcome_nonexistent  0.000000
44     contact_telephone  0.000000

[63 rows x 2 columns]